hpat/_distributed.cpp

```diff
@@ -470,6 +480,130 @@ void permutation_int(int64_t* output, int n)
+}
+
+// Returns an |index_array| which would sort the array |v| of size |len| when
+// applied to it.  Identical to numpy.argsort.
+template<class T>
+std::vector<size_t> arg_sort(T *v, int64_t len) {
+    std::vector<size_t> index_array(len);
+    std::iota(index_array.begin(), index_array.end(), 0);
+    std::sort(index_array.begin(), index_array.end(),
+              [&v](size_t i1, size_t i2) {return v[i1] < v[i2];});
+    return index_array;
+}
+
+// |v| is an array of elements of size |elem_size|.  This function swaps
+// elements located at indices |i1| and |i2|.
+void elem_swap(unsigned char *v, int64_t elem_size, size_t i1, size_t i2)
+{
+    std::vector<unsigned char> tmp(elem_size);
+    auto i1_offset = v + i1 * elem_size;
+    auto i2_offset = v + i2 * elem_size;
+    std::copy(i1_offset, i1_offset + elem_size, tmp.data());
+    std::copy(i2_offset, i2_offset + elem_size, i1_offset);
+    std::copy(std::begin(tmp), std::end(tmp), i2_offset);
+}
+
+// Applies the permutation represented by |p| to the array |v| whose elements
+// are of size |elem_size|.
+void apply_permutation(unsigned char *v, int64_t elem_size,
+                       std::vector<size_t>& p)
+{
+    std::vector<unsigned char> tmp(elem_size);
+    for (size_t i = 0; i < p.size(); ++i) {
+        auto current = i;
+        while (i != p[current]) {
+            auto next = p[current];
+            elem_swap(v, elem_size, next, current);
+            p[current] = current;
+            current = next;
+        }
+        p[current] = current;
+    }
+}
+
+// Applies the permutation represented by |p| of size |p_len| to the array |rhs|
+// of elements of size |elem_size| and stores the result in |lhs|.
+void permutation_array_index(unsigned char *lhs, int64_t len, int64_t elem_size,
+                             unsigned char *rhs, int64_t *p, int64_t p_len)
+{
+    if (len != p_len) {
+        std::cerr << "Array length and permutation index length should match!\n";
+        return;
+    }
+
+    auto num_ranks = hpat_dist_get_size();
+    auto rank = hpat_dist_get_rank();
+    auto dest_ranks = find_dest_ranks(rank, num_ranks, p, p_len);
+    auto send_counts = find_send_counts(dest_ranks, num_ranks, elem_size);
+    auto send_disps = find_disps(send_counts);
+    auto recv_counts = find_recv_counts(rank, num_ranks, p, p_len, elem_size);
+    auto recv_disps = find_disps(recv_counts);
+
+    auto offsets = send_disps;
+    std::vector<unsigned char> send_buf(dest_ranks.size() * elem_size);
+    for (size_t i = 0; i < dest_ranks.size(); ++i) {
+        auto send_buf_offset = offsets[dest_ranks[i]]++ * elem_size;
+        auto *send_buf_begin = send_buf.data() + send_buf_offset;
+        auto *rhs_begin = rhs + i * elem_size;
+        std::copy(rhs_begin, rhs_begin + elem_size, send_buf_begin);
+    }
+
+    MPI_Alltoallv(send_buf.data(), send_counts.data(), send_disps.data(),
+                  MPI_INT64_T, lhs, recv_counts.data(), recv_disps.data(),
```

Comment:
Needs MPI data type of size `elem_size`.

Reply:
Fixed.